R/PipeOpTrafo.R

```diff
@@ -45,168 +27,200 @@
+      # should be overloaded
+      list()
     },
 
-    set_state = function(task) {
-      # set the state, can be list(NULL)
-      # this will internally be called a single time during .train prior to train_target
+    .transform = function(task, phase) {
+      # return a modified task, should make use of private$.update_target
       stop("Abstract.")
     },
 
-    train_target = function(task) {
-      # return a modified task, make use of private$.update_target
-      stop("Abstract.")
+    .train_invert = function(task) {
+      # return a predict_phase_state object (can be anything)
+      list(truth = task$truth())
     },
 
-    train_invert = function(task) {
-      # return a predict_phase_control object (can be anything)
+    .invert = function(prediction, predict_phase_state) {
+      # function that inverts the predictions and returns a Prediction object
       stop("Abstract.")
     },
 
-    inverter = function(prediction, predict_phase_control) {
-      # function that inverts the predictions and returns a Prediction object
-      stop("Abstract.")
-    }
-  ),
-  private = list(
     .train = function(inputs) {
-      intask = inputs[[1]]$clone(deep = TRUE)
-      self$state = self$set_state(intask)
-      intask = self$train_target(intask)
+      intask = inputs[[1L]]$clone(deep = TRUE)
+      self$state = private$.get_state(intask)
+      intask = private$.transform(intask, "train")
       list(NULL, intask)
     },
 
     .predict = function(inputs) {
-      intask = inputs[[1]]$clone(deep = TRUE)
-      intask = self$train_target(intask)
-      predict_phase_control = self$train_invert(inputs[[1L]])
+      intask = inputs[[1L]]$clone(deep = TRUE)
+      predict_phase_state = private$.train_invert(intask)
+      intask = private$.transform(intask, "predict")
       list(
-        fun = private$.invert_help(predict_phase_control),
+        fun = private$.invert_help(predict_phase_state),
         task = intask
       )
     },
 
-    # updates the target of a task and also the task_type (if needed), uses convert_task
-    .update_target = function(task, new_target, new_type = NULL, ...) {
-      convert_task(task, new_target = new_target, new_type = new_type, drop_original_target = TRUE, ...)
-    },
-
-    .invert_help = function(predict_phase_control) {
+    .invert_help = function(predict_phase_state) {
       # provides a helper function for the inversion
-      # PipeOpTargetInverter will later apply this function during prediction
+      # PipeOpTargetInvert will later apply this function during prediction
       function(inputs) {
         assert_list(inputs, len = 1L, types = "Prediction")
-        self$inverter(inputs[[1L]], predict_phase_control)
+        list(private$.invert(inputs[[1L]], predict_phase_state))
       }
+    },
+
+    # updates the target of a task and also the task_type (if needed), uses convert_task
+    .update_target = function(task, new_target, new_type = NULL, ...) {
```

Comment:
When I thought up the ".update_target" function I thought it would contain all the business code, but that is now in convert_task. I think a better solution now would be to just have users call `convert_task` inside their `.transform()` function, and to remove the `private$.update_target` thing altogether.

Reply:
Makes sense, I will fix this!